lib/gomod-parser.ts

```diff
@@ -0,0 +1,150 @@
+import { InvalidUserInputError } from './errors';
+import {
+  GoMod,
+  ModuleAndVersion,
+  Require,
+  ModulePseudoVersion,
+  ModuleExactVersion,
+  ModuleVersion,
+  ModuleAndMaybeVersion,
+} from './types';
+
+// go.mod file format:
+// https://tip.golang.org/cmd/go/#hdr-The_go_mod_file
+
+// See https://tip.golang.org/cmd/go/#hdr-Pseudo_versions
+// Subgroups: baseVersion, suffix, timestamp, hash
+const rePseudoVersion = /(v\d+\.\d+\.\d+)-(.*?)(\d{14})-([0-9a-f]{12})/;
+const reIndirect = /\/\/ indirect/;
+const reExactVersion = /^(.*?)(\+incompatible)?$/;
+
+type Statement = 'module' | 'go' | 'require' | 'exclude'| 'replace';
+
+const reStatementWord = /^(module|go|require|replace|exclude) (\(?)/;
+
+const reLineWithComments = /^(.*?)(\/\/.*)?$/;
+
+function parseModuleAndMaybeVersion(lineRemainder: string): ModuleAndMaybeVersion {
+  const [moduleName, versionString] = lineRemainder.trim().split(' ');
+  if (!versionString) {
+    return {moduleName};
+  }
+  return {moduleName, version: parseVersion(versionString)};
+}
+
+function parseModuleAndVersion(lineRemainder: string): ModuleAndVersion {
+  const [moduleName, versionString] = lineRemainder.trim().split(' ');
+  if (!moduleName || !versionString) {
+    throw new Error(`could not split "${lineRemainder}" into moduleName and version`);
+  }
+  return {moduleName, version: parseVersion(versionString)};
+}
+
+function parseVersion(versionString: string): ModuleVersion {
+  const maybeRegexMatch = rePseudoVersion.exec(versionString);
+  if (maybeRegexMatch) {
+    const [baseVersion, suffix, timestamp, hash] = maybeRegexMatch.slice(1);
+    return {baseVersion, suffix, timestamp, hash};
+  } else {
+    // No pseudo version recognized, assuming the provided version string is exact
+    const [exactVersion, incompatibleStr] = reExactVersion.exec(versionString)!.slice(1);
+    return {exactVersion, incompatible: !!incompatibleStr};
+  }
+}
+
+export function parseGoMod(goModStr: string): GoMod {
+  try {
+    const lines = goModStr.split('\n');
+
+    const res: GoMod = {
+      moduleName: '',
+      requires: [],
+      replaces: [],
+      excludes: [],
+    };
+
+    let inMultilineStatement = false;
+    let statement: Statement | null = null;
+
+    for (let line of lines) {
+      line = line.trim();
+
+      // Find out whether we are in an one-line or multiline statement
+      if (!inMultilineStatement) {
+        const maybeStatement = reStatementWord.exec(line);
+        if (maybeStatement) {
+          statement = maybeStatement[1] as Statement;
+          inMultilineStatement = !!maybeStatement[2]; // whether we found "("
+          if (!inMultilineStatement) {
+            line = line.substr(statement.length + 1).trim();
+          } else {
```

Comment:
instead of these continues, can you put an else on line 96 - 127?

Reply:
Converted to immutable state transitions, which are a bit more verbose but less error-prone.